spec/core/instances_spec.rb

```diff
@@ -436,31 +449,79 @@
   end
 
   describe '#add_instance' do
-    it 'adds an instance from given values to the Instances object' do
-      data = [:sunny, 70, 80, 'TRUE', :yes]
-      subject.add_instance(data)
 
-      expect(subject.instances.last.to_s).to eq data.join(',')
-    end
+    context 'when passing an array of attribute values' do
+      it 'adds an instance from given values to the Instances object' do
+        data = [:sunny, 70, 80, 'TRUE', :yes]
+        subject.add_instance(data)
+
+        expect(subject.instances.last.to_s).to eq data.join(',')
+      end
 
-    it 'adds a given instance to the Instances object' do
-      data = subject.first
-      subject.add_instance(data)
+      it 'adds a given instance to the Instances object' do
+        data = subject.first
+        subject.add_instance(data)
 
-      expect(subject.instances.last.to_s).to eq data.to_s
-    end
+        expect(subject.instances.last.to_s).to eq data.to_s
+      end
 
-    it 'adds a given instance with only missing values' do
-      data = Weka::Core::DenseInstance.new(subject.size)
-      subject.add_instance(data)
-      expect(subject.instances.last.to_s).to eq data.to_s
+      it 'adds a given instance with only missing values' do
+        data = Weka::Core::DenseInstance.new(subject.size)
+        subject.add_instance(data)
+        expect(subject.instances.last.to_s).to eq data.to_s
+      end
+
+      it 'adds a given instance with partly missing values' do
+        data = [:sunny, 70, nil, '?', Float::NAN]
+        subject.add_instance(data)
+
+        expect(subject.instances.last.to_s).to eq 'sunny,70,?,?,?'
+      end
     end
 
-    it 'adds a given instance with partly missing values' do
-      data = [:sunny, 70, nil, '?', Float::NAN]
-      subject.add_instance(data)
+    context 'when passing a hash of attribute values' do
+      let(:data) do
+        {
+          outlook: :sunny,
+          temperature: 70,
+          humidity: humidity_value,
+          windy: windy_value,
+          play: play_value
+        }
+      end
+
+      let(:humidity_value)  { 80 }
+      let(:windy_value)     { 'TRUE' }
+      let(:play_value)      { :yes }
```

Comment:
It's kind of weird to see the actual values be defined after their usage. Should the lets be defined before `let(:data)...`?